proposed/resource-limit.md

```diff
@@ -0,0 +1,559 @@
+
+#### Resource limit implementations
+
+In addition to defining the abstractions, we would also like to ship some default implementations of common types of resources limiters. Details on how the functionality described here are achieved can be found in the [implementation section](#concrete-implementations).
+
+1. Fixed window rate limiter
+
+This is the simplest type of rate limiter. It counts the number of resources acquired in a time span. If the resource limit is reached, additional calls to `TryAcquire` will return `false` and additional calls to `AcquireAsync` will wait if queuing is enabled until the queue limit. The resource is refreshed when the current window has elapsed.
+
+2. Sliding window rate limiter
+
+This is another type of rate limiter. Counts for the number of resources acquired in a time segment is kept. A window is defined as multiple time segments and the resource is considered to be exhausted if the total across the segments in the window is exceeded. If the resource is exhausted, additional calls to `TryAcquire` will return `false` and additional calls to `AcquireAsync` will wait if queuing is enabled until the queue limit. The window is updated when the current time segment has elapsed.
+
+As an example consider the following example:
+
+```
+Settings: window time span = 3s, number of segments per window = 3, resource limit = 10
+
+Time:   |1s |2s |3s |4s |
+Count:  [3] [4] [3] [1]
+        | Window 1  |
+            | Window 2  |
+```
+
+In the above example, resource is exhaused in Window 1 since the total count is 10 and no additional resources can be acquired in the 3rd second. On the 4th second, Window 2 now has a count 8 and can accept additional resource acquisition requests.
+
+3. Token bucket rate limiter
+
+This rate limiter uses the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket). The functionality will be similar to the two implementation described above but will be configured differently.
+
+4. Concurrency limiter
+
+This limiter keeps track of the number of concurrent resources in use. Conceptually, it will function in a similar way to a Semaphore/SemaphoreSlim but adapted to the proposed API. Like the rate limits above, it will also keep a queue of pending acquisition requests if configured to do so.
+
+#### Resource limit adoption
+
+Two BCL types where adoption of the resource limiters are beneficial and relatively simple are:
+
+- System.Threading.Channels
+- System.IO.Pipelines
+
+### Non-Goals
+
+#### HttpClient, Streams, Sockets
+
+Although we anticipate that there will be applications in many .NET areas and components such as HttpClient, Streams and Sockets, these scenarios will be explored but not committed for adoption in .NET 6 since they may be more complex. The adoption of the resource limiting APIs in these areas can happen independently of this proposal in the future.
```

Comment:
It would be useful to design the APIs for HttpClient, Streams and Sockets, even if we are not planning the implement them in .NET 6.

Reply:
This is important to decide where it needs to be in the stack. The key question - does it need to be in CoreLib?

Reply:
@JunTaoLuo for Streams, we should show a hypothetical ResourceLimitedStream decorator (with settings for read and write). For Sockets I'm not sure we have an obvious place to put it at that layer since we don't rate limit at that level in the stack. For HttpClient there are places in the implementation we might want to use this for (HTTP2 flow control for example?).

Reply:
> does it need to be in CoreLib?

We can start outside of CoreLib, and even outside of the shared framework, and move it down in the future. We've done that with other APIs.

Reply:
I don't think it needs to be in CoreLib, but it needs to be low enough in the stack for things like Pipelines and channels to be able to use them. I don't know where that is right now. Maybe a new assembly.

Reply:
System.Threading.ResourceLimits?

Reply:
> Maybe a new assembly

This was my thinking - a new library.

> but it needs to be low enough in the stack for things like Pipelines and channels to be able to use them

Pipelines isn't in the shared framework today. Channels is. So if we needed Channels to depend on this new library, it would need to be in the shared framework.

Channels is extensible though, so if we really wanted to avoid putting this new library in the shared framework, I think we design a way to allow Channel integration, but without having Channels depend on the new library.

Reply:
I agree that it feels like a higher level abstraction (from the runtime point of view), and thus can start outside the core framework or outside CoreLib at least.

From the examples I have seen so far, we will be typically creating wrappers of existing types that are resource limited. These wrappers do not need to live in the next to the type that they are wrapping. For example, the channels one does not really have to live in the System.Threading.Channels assembly.

If you agree, I would suggest to reword "Shared abstractions and implementation for the entire stack" paragraph. This paragraph makes it sound like that the core types have to be tightly coupled with the resource limiters. I have not seen a good reason for that so far.

Reply:
I see this like an interface for SemaphoreSlim, I think it belong at that layer of the stack but I'm willing to not put this in CorLib until we're happy with the design. See my comment here about bridge packages https://github.com/dotnet/designs/pull/215#discussion_r631915858.

The argument being made here breaks down when it wants to be used inside of HttpClient (which is also in the shared framework).

Reply:
So far I've been prototyping with the assumption it will be a new library in the shared framework, coupling tightly with types such as Channels. But I'm willing to adjust this as necessary based on the discussions here.